src/classes/syncer/WebResourceSyncer.ts

```diff
@@ -0,0 +1,72 @@
+import * as vscode from 'vscode';
+import { WebResoureceSyncerResponse, ListWebResourcesInSolutionAction } from '../../models/WebResourceSyncerResponse';
+import util = require('util');
+
+
+
+export class WebResourceSyncer {
+	_exePath: string;
+	_execFile: Function;
+
+
+	constructor(exePath: string) {
+		this._exePath = exePath;
+		this._execFile = util.promisify(require('child_process').execFile);
+	}
+
+	async retreiveWebResourcesInSolution(solutionName: string): Promise<{
+		name: string;
+		id: string;
+	}[]> {
+		const args = ['--solution', solutionName, '--listWebResources',];
+
+		const procResult = await this._execFile(this._exePath, args, {
+			shell: true,
+			windowsHide: true,
+		});
+
+		let response: WebResoureceSyncerResponse = JSON.parse(procResult.stdout);
+
+		if (response.dryRun) {
+			await vscode.window.showInformationMessage("Dry run successful");
+			return Promise.resolve([]);
+		}
+
+		for (let action of response.actionList) {
+			let string = `Stage: ${action.actionName}. Successful: ${action.successful}. ` + (action.successful ? "" : `Error message: ${action.errorMessage}`);
+			vscode.window.showInformationMessage(string, "Reun this stage?");
+
+			if (action.actionName === "ListWebResourcesInSolution") {
+				let listAction = action as ListWebResourcesInSolutionAction;
+
+				return listAction.webResources;
+			}
+		}
+		throw new Error("No ListWebResourceAction found?");
+	}
+
+	async uploadFile(solutionName: string, path: string, publish: boolean = false) {
+		const args = ['--filePath', path, '--solution', solutionName, '--updateIfExists',];
+
+		if (publish) {
+			args.push('--publish');
+		}
+
+		const procResult = await this._execFile(this._exePath, args, {
+			shell: true,
+			windowsHide: true,
+		});
+
+		let response: WebResoureceSyncerResponse = JSON.parse(procResult.stdout);
```

Comment:
This may be something to think about for the future. Is this really the best way to pass "messages" from the extension to the syncer? Probably good enough for an MVP - but if this thing grows *big* then we may need to think about going back to your idea of making the syncer like a WEB API that just listens over some *obscure* port for http connections. Cause then we could get that nice request/response pattern.

Reply:
I'm scared of listening over ports 😵‍💫

What's wrong with passing messages this way?

Reply:
> I'm scared of listening over ports 😵‍💫
> 
> What's wrong with passing messages this way?

Yea good points - I'll address them 1 by 1.

1: I totally agree that listening on an http port is a bit invasive on the user's machine. But - somehow I think there should be a way we can just listen on a "process port" or something like that. Cause that's how most LSP's do it. Like the [Rust Language Server](https://github.com/rust-lang/rls) for example. It's a process that just sits there and waits for clients to connect to it with a Rust source file and it can respond and say "yea dog that syntax looks good to me". And it's some kind of "process only server" I think. Doesn't accept connections over http. At least that's my current understanding of how LSPs do it. 😛 

2: I don't think there's something inherently wrong with doing it this way, Now you got me thinking about as much decoupling as we can hehe. So my thought is: the general pattern as it stands today is the `Extension` sends a sequence of instructions to the `Syncer`:

- Open connection
- Upload files A,B,C,D,E to solution X
- Publish files: D,E,F
- Refresh the entire solution list and get me all of the updated files

What happens if one of the intermediate steps fails? Should the `Syncer` just be like "Ok, step 2 failed, I'll stop execution and just return the response object". Or should the `Syncer` be like "Ok, step 2 failed, but I can perform step 4 without needing step 2, I'll do my best to keep going but still mark step 2 as failed" 

My gut instinct is that it might place too much "responsibility" on the `Syncer` to "have an opinion" of how to handle such cases, whereas maybe it makes more sense to just have a "simpler" request/response pattern, where it's up to the extension to decide when/where it want's to execute steps.

IDK - also maybe I'm overthinking this 😄 . Definitely doesn't doesn't stop these changes from going through though - this PR looks beautiful to me.

Reply:
1. I honestly wouldn't know how to do this, so you'd have to teach me and let me know what documentation to read. So, I'm open to it, but I am also hesitant because of how long it took us to finally get to the executable idea. I also worry about security issues when opening up ports... I don't know enough to really say if these concerns are valid. Do you? If you can convince me that we won't invite hackers into people's computers & that it won't take another 6 months, I'm down.

2. I'm convinced that we can solve this problem without switching our structure. You're right that if there is a failure, the program should stop. A half-dead program is deadlier than a dead one. I also agree with you, though, that `syncer` should know nothing about the sequence of events that is happening.

This is similar to the problem described on page 154 of The Pragmatic Programmer.  If we abstract those steps enough, what really have is a linear chain of events, or a pipeline (even though no transformations to our actual data is happening). Some suggestions the authors make are: 

- **Handle it in the transformation** Have a wrapper passed to each function that represents success or failure, and when a function receives it as an input parameter, that function should do nothing if it receives failure, or continue if it receives success. (I don't like this one too much because it assumes that `syncer` has knowledge of the pipeline.)
- **Handle it in the pipeline** Our pipeline itself reads the input from the last function, and only on success does it continue. I believe this is the best way to go, and I believe that addresses all your concerns. If it doesn't, let me know. I'm enjoying this exercise

In both cases we must define a success/failure wrapper that will be passed around. What do you think? My brain is going 1,000,000mph and I can see the 5th dimension.

Reply:
Hehe don't bust a belt trying to over work that smart :brain:  of yours! :pleading_face: 

Ok - but I think I'm tracking with you - and I agree that it should just stop on first failure. 

I'm not so sure how we could implement the concept of "passing a function" to a distinct process without setting up a JSON RPC server. (Ironically, I **think** this is exactly the sort of thing JSON RPC was developed for)

So - simplest version of solution for now: I'll add a task in ADO and assign it to myself to update `Syncer` to just halt after a failure. And then I'll add another task (with the info from this convo) for "Sort out Communication Contract Between `Syncer` and `Extension`" and we'll get to that one down the road?

Reply:
Sweet! We can talk about it whenever we next meet

Reply:
Right on - do you want me to merge this PR? Or are you gonna do it when you're good and ready : )

Reply:
Not home right now, feel free to merge if you can, or I can do it tomorrow